greedy.py

```diff
@@ -1,25 +1,135 @@
+            (at_coord[0], at_coord[1] + 1),
+            (at_coord[0], at_coord[1] - 1),
+            (at_coord[0] + 1, at_coord[1]),
+            (at_coord[0] - 1, at_coord[1])]
+
+        def inside_board(step):
+            return (0 <= step[0] < 8 and 0 <= step[1] < 8)
+
+        def criss_cross(at_coord, step):
+            def piece_moved():
+                return (step[0] < at_coord[0] or step[1] < at_coord[1])
+            def ally_located_at_step():
+                return board[step[0]][step[1]] == ant_team
+            return (at_coord in teammate_positions and piece_moved() and ally_located_at_step())
+
+        # remove steps that kill teammate ants (remove friendly fire) and off board steps
+        # Also prevents criss-cross steps (when ant1 chooses to move to ant2's position
+        # and ant2 chooses to move to ant1's position)
+        possible_steps = [step for step in all_steps if (step not in teammate_positions
+                            and inside_board(step) and not criss_cross(at_coord, step))]
+        return min(possible_steps, key=lambda x:min_max_dist(x, min_coord, max_coord))
+
+    # Define actions per ant
     actions = []
-    
-    # TODO: Fix this (I am doing random action for each ant)
-    # NOTE: An invalid action will appear to do nothing (e.g. moving ant out of board)
+    teammate_positions = set()
     for i in range(ac.ROW):
         for j in range(ac.COL):
             if board[i][j] == ant_team:
-                rand_action = random.randint(0, len(ac.Action)-1)
-                actions.append((i, j, ac.Action(rand_action)))
+                closest_food = closest((i, j), food_locations)
+                closest_enemy = closest((i, j), enemies)
+                if dist((i,j), closest_enemy) == 1 and closest_enemy not in teammate_positions:
+                    step = closest_enemy
+                else:
+                    step = min_risk_max_reward_step((i,j), closest_food, closest_enemy, board, teammate_positions)
+                action = get_action((i,j), step)
+                actions.append((i, j, action))
+                teammate_positions.add(step)
+
+    # Orders action list to allow ants to move safely
+    # Ants moving to an empty/enemy/food position will move first
+    def order_actions(actions):
+        def new_pos(pos, action):
+            if action == ac.Action.LEFT:
+                return (pos[0], pos[1]-1)
+            elif action == ac.Action.RIGHT:
+                return (pos[0], pos[1]+1)
+            elif action == ac.Action.UP:
+                return (pos[0]-1, pos[1])
+            elif action == ac.Action.DOWN:
+                return (pos[0]+1, pos[1])
+            else:
+                return pos
+        def safe_action(new_position, board):
+            if board[new_position[0]][new_position[1]] != ant_team:
+                return True
+            else:
+                return False
+        tempoarary_board = copy.deepcopy(board)
```

Comment:
Just a little typo here with `temporary`. Also I've been using `new_board = [row[:] for row in old_board]` instead of `copy.deepcopy` because it's faster apparently and prevents the `import copy`. Probably not significant, but it's nice if the codebase uses the same thing.